pkg/kv/kvserver/kvflowcontrol/replica_rac2/admission.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package replica_rac2
+
+import (
+	"cmp"
+	"slices"
+
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
+	"github.com/cockroachdb/errors"
+)
+
+// lowPriOverrideState records which raft log entries have their priority
+// overridden to be raftpb.LowPri. Used at follower replicas.
+//
+// The lifetime of a particular log index begins when the entry is appended
+// to the in-memory log, and ends when the entry is sent to storage with the
+// priority that is extracted from this struct. The same log index can
+// reappear in this struct multiple times, when a new raft leader overwrites a
+// suffix of the log proposed by the previous leader, or if the existing
+// leader retransmits an entry.
+//
+// sideChannelForLowPriOverride is called when stepping the MsgApps in a
+// RaftMessageRequest, and getEffectivePriority is called from within
+// handleRaftReady when processing the corresponding MsgStorageAppend.
+//
+// Maintaining accuracy in the presence of an existing leader doing retransmits
+// is tricky, since the caller does not know which MsgApps were accepted by
+// Raft during a step. We can have two kinds of situations:
+//
+//   - [C1] e10 was stepped, but not accepted, since e9 was missing. e10 is
+//     stepped successfully later.
```

Comment:
The MsgApp can also be rejected if sent from a leader with an older term than ours.